posthog/middleware.py

```diff
@@ -629,3 +630,39 @@ def process_response(self, request, response):
             )
 
         return response
+
+
+class AutoLogoutImpersonateMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request: HttpRequest):
+        if not is_impersonated_session(request):
+            return self.get_response(request)
+
+        init_time = request.session.setdefault(settings.IMPERSONATION_SESSION_KEY, time.time())
+        session_is_expired = time.time() - init_time > settings.IMPERSONATION_TIMEOUT_SECONDS
+
+        if session_is_expired:
+            # This works but has an issue in that api requests will all stop working...
+
+            if request.path.startswith("/api/"):
+                # If we respond with a 401, the frontend will check and redirect to logout
+                return HttpResponse(
+                    "Impersonation session has expired. Please log in again.",
+                    status=401,
+                )
+            elif not request.path.startswith("/logout"):
+                # Any other endpoint will redirect to the logout page (this could be doc level calls or other things like static assets)
+                return redirect("/logout/")
+            else:
+                # And finally we restore the login and redirect to the admin page
+                restore_original_login(request)
+                return redirect("/admin/")
```

Comment:
What is left here in this else branch? Wouldn't this just match the `/logout` route? And for logout, couldn't we just let the request run through (it would stop impersonating, right)?